### PR TITLE
[HiCache & HybridModel] nixl hicache backend support hybrid models

### PR DESCRIPTION
**Pull request by ziruiliu**
## Motivation

Previously, PR #25538 was created to add support for the HiCache NIXL backend to store hybrid models. However, HiCache NIXL backend has been refactored.
After the HiCache NIXL backend refactor, the previous hybrid-storage change became stale because NIXL now has its own registry, pre-registered host memory path, O_DIRECT alignment handling, and persistent bounce-buffer flow. This PR reworks NIXL hybrid HiCache support on top of the new backend structure, so hybrid models like Qwen3.5/Mamba and DSA-style side pools can use NIXL as the HiCacheHybrid controller storage backend.

## Modifications

- Implement real `register_mem_host_pool_v2` in `HiCacheNixl`.
  It now registers each side-pool with NIXL using a per-pool context, instead of only saving the pool reference. If a side pool supports zero-copy and passes alignment check, its host buffers are pre-registered with the NIXL agent.

- Add zero-copy selection for aligned side pools.
  NIXL checks `get_hybrid_pool_buffer()`, `get_page_buffer_meta()`, and `is_stride_page_aligned()` before enabling side-pool zero-copy. This directly addresses the feedback that zero-copy should be guarded by alignment and real NIXL registration.

- Add persistent per-pool GET/SET bounce buffers for fallback path.
  When zero-copy is not safe, each registered side-pool gets its own persistent `bounce_get` and `bounce_set` buffers. This avoids sharing one bounce buffer across pools or directions, so tensor bookkeeping stays simple and follows the existing GET/SET thread model.

- Make `batch_get_v2` and `batch_set_v2` use the host-pool-specific transfer context.
  v2 transfers now either use registered zero-copy buffer metadata or the pool-specific bounce buffers. They also emit debug transfer stats like v1, e.g. `batch_set_v2[mamba] transferred: ...`, making side-pool storage traffic easier to verify.

- Keep `batch_exists_v2` simple and pool-aware.
  It formats the sidecar object names and queries NIXL storage similarly to v1. For multi-object side pools, it folds object-level existence results back to page-level hit counts and still respects `ALL_PAGES` and `TRAILING_PAGES` policies.

- Add consistent sidecar key naming.
  Sidecar objects use suffixes after the normal NIXL base key, for example `_mamba_temporal`, `_mamba_conv_0`, and pool-specific `_k` / `_v` names when a side pool expands into multiple objects.

- Add `is_stride_page_aligned()` support for side-pool host classes.
  `MambaPoolHost`, `DeepSeekV4PagedHostPool`, `DeepSeekV4StateHostPool`, `DSAIndexerPoolHost`, and `HostPoolGroup` now expose alignment checks needed by the NIXL zero-copy decision.

- Add unit coverage for NIXL hybrid v2 behavior.
  Tests cover zero-copy side-pool registration, bounce-buffer fallback registration, expanded Mamba sidecar keys, and v2 GET/SET behavior through the per-pool transfer path.

- Update NIXL README and add an e2e validation script.
  The README now includes the Qwen3.5 hybrid NIXL serve command and the expected sidecar file naming. The local script `~/TestEnv/nixl_hicache_hybrid_e2e.py` starts SGLang with NIXL HiCache storage, sends a long request, and checks that NIXL is selected by the hybrid controller and that Mamba sidecar files are written.

After this change, cache objects are distributed in subdirectories (for POSIX backend) like:
```
c41f2e1b1bdad16596d5901ce236ce6a32bd8e0d8d6cfcdda5c3924342cf92ad_Qwen-Qwen3.5-9B_0_2_v
c41f2e1b1bdad16596d5901ce236ce6a32bd8e0d8d6cfcdda5c3924342cf92ad_Qwen-Qwen3.5-9B_0_2_mamba_conv_0
c41f2e1b1bdad16596d5901ce236ce6a32bd8e0d8d6cfcdda5c3924342cf92ad_Qwen-Qwen3.5-9B_1_2_mamba_temporal
c41f2e1b1bdad16596d5901ce236ce6a32bd8e0d8d6cfcdda5c3924342cf92ad_Qwen-Qwen3.5-9B_1_2_v
c41f2e1b1bdad16596d5901ce236ce6a32bd8e0d8d6cfcdda5c3924342cf92ad_Qwen-Qwen3.5-9B_1_2_mamba_conv_0
c41f2e1b1bdad16596d5901ce236ce6a32bd8e0d8d6cfcdda5c3924342cf92ad_Qwen-Qwen3.5-9B_1_2_k
c41f2e1b1bdad16596d5901ce236ce6a32bd8e0d8d6cfcdda5c3924342cf92ad_Qwen-Qwen3.5-9B_0_2_k
c41f2e1b1bdad16596d5901ce236ce6a32bd8e0d8d6cfcdda5c3924342cf92ad_Qwen-Qwen3.5-9B_0_2_mamba_temporal
```

## Accuracy Tests

will append later

## Speed Tests and Profiling

will append later

## Checklist

- [X] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [X] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [X] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #29244425711](https://github.com/sgl-project/sglang/actions/runs/29244425711)
Latest PR Test (Extra): :x: [Run #29244425701](https://github.com/sgl-project/sglang/actions/runs/29244425701)

### CODE REVIEW

**Comment by ziruiliu** [comment]
@lluki  This is the new PR that enable hybrid models in NIXL backend, after NIXL's refactor. All your concerns/feedbacks are addressed. Please take a look.

File: python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py [hunk 1]
```diff
@@ -620,3 +838,135 @@ def batch_set_v1(
...
+        extra_info: Optional[HiCacheStorageExtraInfo] = None,
+    ) -> PoolTransferResult:
+        kv_pages = self.batch_exists(keys, extra_info)
+        hit_count: dict = {PoolName.KV: kv_pages} if kv_pages else {}
+        final_pages = kv_pages
+
+        for transfer in pool_transfers or []:
+            if final_pages == 0:
+                break
+            if transfer.name not in getattr(self, "registered_pools", {}):
```
Comment (lluki) [inline python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py]: Its from the base class, but it would be better to just make `registered_pools` be initialized in init (in the super class `__init__` i mean) se we can just use `self.registered_pools`

File: python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py [hunk 2]
```diff
@@ -322,6 +381,170 @@ def register_mem_pool_host(self, mem_pool_host: HostKVCache):
...
+            logger.error(
+                "HiCacheNixl: hybrid pool %s key/meta mismatch: keys=%s ptrs=%s",
+                transfer.name,
+                len(key_strs),
+                len(ptr_list),
+            )
+            return [], [], 0
+        return key_strs, list(zip(ptr_list, size_list)), key_multiplier
+
+    def _prepare_pool_transfer(
```
Comment (lluki) [inline python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py:524]: can we get a docstring here what this function does and what is returned

File: python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py
Comment (lluki) [inline python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py:823]: Would it be possible to call `batch_get_v2` and avoid some duplication? (Same for the _set case)

Comment (lluki) [reply python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py:823]: I know this would invalidate existing cache files because the filenames are formatted differently, but given this is just a cache, i think it's OK to invalidate the cache on SGLang upgrade.

File: test/registered/unit/mem_cache/test_hicache_nixl_storage.py [hunk 3]
```diff
@@ -494,6 +548,83 @@ def test_batch_exists_zero_copy_mha_uses_two_key_denominator(self):
...
+                    name=PoolName.MAMBA,
+                    keys=["p0"],
+                    host_indices=torch.tensor([0], dtype=torch.int64),
+                )
+            ]
+        )
+
+        self.assertEqual(results[PoolName.MAMBA], [True])
+        self.assertTrue(torch.all(pool.get_data_page(0) == 3))
+
```
Comment (lluki) [inline test/registered/unit/mem_cache/test_hicache_nixl_storage.py:644]: Can we add a simple set-then-get testcase using the v2 API ? And ensure we correctly distinguish on the PoolName? For example `set(Pool1, key1, data1); set(Pool2, key1, data2); assert(get(Pool1,key1) == data1); assert(get(Pool2,key1) == data2)`

**Review by lluki — CHANGES_REQUESTED** [review]
Thanks for the contribution!

For a future PR, we should consider also batching together all requests in `batch_(g|s)et_v2` to have just one NIXL xfer invocation instead of looping over the transfers in python.

File: python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py (hunk 1 above)
Comment (ziruiliu) [reply python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py]: Do you mean HiCacheStorage by super class? Touching HiCacheStorage would affect other backends. My idea here is to initialize pools in HiCacheNixl.__init__:
self.registered_pools: dict[PoolName, HostKVCache] = {}

I will add it in my next commit

File: python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py (hunk 2 above)
Comment (ziruiliu) [reply python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py:524]: Added in my next commit

File: python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py
Comment (ziruiliu) [reply python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py:823]: batch_set_v1/get_v1 operate on the legacy KV pool registration, while batch_set_v2/get_v2 operate on registered PoolTransfer side pools. 
Unifying them is possible by registering the default KV pool into the v2 context and making v1 wrappers call v2, but that is a separate refactor with some risk to the existing KV-only path.
At this time, I would prefer to keep this PR scoped to hybrid side-pool support.

File: test/registered/unit/mem_cache/test_hicache_nixl_storage.py (hunk 3 above)
Comment (ziruiliu) [reply test/registered/unit/mem_cache/test_hicache_nixl_storage.py:644]: ok, I will add a set/get test case in my next commit

File: python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py
Comment (lluki) [reply python/sglang/srt/mem_cache/storage/nixl/hicache_nixl.py:823]: Ok. let's keep it like it is

**Review by lluki — APPROVED** [review]

**Comment by ziruiliu** [comment]
The latest "Fix dsv4 issue" commit  fixes NIXL registration for DeepSeek V4 Flash HiCache.

DeepSeek V4 uses a LogicalHostPool as the KV anchor, where kv_buffer is intentionally None and real cache bytes are stored in sidecar pools. NIXL previously assumed every primary host pool had a real kv_buffer for registration.
The fix treats this logical anchor as a marker-only pool while continuing to transfer real DSV4 sidecar data through the v2 pool APIs.

**Comment by ziruiliu** [comment]
For deepseek v4 flash model, sglang + hicache + nixl + POSIX produces cache objects like:
```
6146050ab782a07eec24014cf1cb93bc2cbe60199bd8016792174bba21790c04_local-dsv4-flash_deepseek_v4_c4_indexer_state
6146050ab782a07eec24014cf1cb93bc2cbe60199bd8016792174bba21790c04_local-dsv4-flash_deepseek_v4_c4
6146050ab782a07eec24014cf1cb93bc2cbe60199bd8016792174bba21790c04_local-dsv4-flash_swa
6146050ab782a07eec24014cf1cb93bc2cbe60199bd8016792174bba21790c04_local-dsv4-flash_deepseek_v4_c4_indexer
6146050ab782a07eec24014cf1cb93bc2cbe60199bd8016792174bba21790c04_local-dsv4-flash_deepseek_v4_c4_state
6146050ab782a07eec24014cf1cb93bc2cbe60199bd8016792174bba21790c04_local-dsv4-flash_deepseek_v4_c128
6146050ab782a07eec24014cf1cb93bc2cbe60199bd8016792174bba21790c04_local-dsv4-flash
```
